src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -93,18 +128,22 @@ private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource
                         }
                     }
 
-                    // Extract the chunk array to yield, then clear out any references in our storage so that we don't keep
-                    // objects alive longer than needed (the caller might clear out elements from the yielded array.)
-                    var chunk = new TSource[i];
-                    Array.Copy(array, 0, chunk, 0, i);
-                    if (RuntimeHelpers.IsReferenceOrContainsReferences<TSource>())
+                    // If the array has the right size return it directly
+                    if (i == array.Length)
+                    {
+                        yield return array;
```

Comment:
This isn't valid.  The method's design involves returning a new array for every chunk, such that someone could do:
```C#
T[][] chunks = source.Chunk(42).ToArray();
```
and get back a valid array of arrays rather than having the same array instance repeated multiple times with overwritten data.

Reply:
Ah, I see you also moved the allocation of the array into the loop.